[CodeCompletion] Explicitly support enum pattern matching

ahoppen opened the pull request:
Pattern matching in Swift can either be expression pattern matching by comparing two instances using the `~=` operator or using enum matching by matching the enum case and its associated types (+ tuple pattern matching, but that’s not relevant here). We currenlty only consider the expression pattern matching case for code completion. To provide enum pattern matching results, we thus need to have a `~=` operator between the code completion token and the match expression

For example, when we are completing

```swift
enum MyEnum {
  case myCase(String)
}

switch x {
case .#^COMPLETE^#
}
```
then we are looking up all overloads of `~=` and try to match it to the call arguments `(<Code Completion Type>, MyEnum)`.
The way we currently get `#^COMPLETE^#` to offer members of `MyEnum`, is that we are trying to make sure that the `~=<T: Equatable>(T, T)` operator defined in the standard library is the best solution even though it has fixes associated with it. For that we need to carefully make sure to ignore other, more favourable overloads of `~=` in `filterSolutions` so that `~=<T: Equatable>(T, T)` has the best score.

This poses several problems:
- If the user defines a custom overload of `~=` that we don't prune when filtering solutions (e.g. `func ~=(pattern: OtherType, value: MyEnum) -> Bool`), it gets a better score than `~=<T: Equatable>(T, T)` and thus we only offer members of `OtherType` instead of members from `MyEnum`
- We are also suggesting static members of `MyEnum`, even though we can't pattern match them due to the lack of the `~=` operator.

If we detect that the completion expression is in a pattern matching position, also suggests all enum members of the matched type. This allows us to remove the hack which deliberately ignores certain overloads of `~=` since we no longer rely on `~=<T: Equatable>(T, T)`. It thus provides correct results in both of the above cases.

Fixes rdar://77263334 [SR-14547]

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -1333,6 +1333,20 @@ ParserResult<Pattern> Parser::parseMatchingPattern(bool isExprBasic) {
   // matching-pattern ::= expr
   // Fall back to expression parsing for ambiguous forms. Name lookup will
   // disambiguate.
+  if (Tok.is(tok::period_prefix) && peekToken().is(tok::code_complete)) {
```
If I understand correctly, I think this doesn't cover `case .foo(.#^HERE^#)`, we don't know the `foo` is an enum element or a function, right? Could you make it work in nested cases too?

Also considering `case .foo(.#^HERE^#)` case, we don't know `foo` is a static function or an enum element. If it's a static function, we can only use `~=` because enum element patterns can't be nested in an expression pattern. Can we differentiate it in `sawSolution()`?

ahoppen replied:
`case .foo(.#^HERE^#)` works because the parser correctly parses the `.foo` as an enum element pattern, so we don’t need to worry about it. I’ll add test cases for it.

rintaro commented on include/swift/IDE/CodeCompletion.h:
```diff
@@ -533,6 +533,9 @@ enum class CompletionKind {
   UnresolvedMember,
+  /// An unresolved member completion in a pattern matching position.
+  /// E.g. \code if case .#^COMPLETE^# = x {} \endcode
+  PatternMatching,
```
I'm not a fan of this naming. Could you include something that imply "after dot" and "without base expr/pattern"? For example `PatternAfterDotWithoutBase` or `UnresolvedDotPattern`?

ahoppen replied:
Renamed to `UnresolvedDotPattern`.

rintaro commented on lib/Sema/TypeCheckCodeCompletion.cpp:
```diff
@@ -1309,6 +1281,42 @@ sawSolution(const constraints::Solution &S) {
+  addTypeToResultVectorIfNecessary(Results.ExprPatternTypes,
+                                   CompletionExprType);
+
+  if (auto MatchVar = PatternMatching->getMatchVar()) {
```
Can't we derive `MatchVar` or `MatchingVarType` from `Solution`? If not, could you make it so?
If we can do that, I think we don't need to add a dedicated completion kind. Instead, we can detect it in `UnresolvedMemberTypeCheckCompletionCallback::sawSolution()`

ahoppen replied:
Not easily, I think. We could maybe search through the constraint locators for one that declares a variable named `$match`, but that feels awfully ad-hoc to me and potentially fragile if `$match` somehow leaks to nested scopes.

Also note that we only offer enum cases for the type of the match var, so `UnresolvedMemberTypeCheckCompletionCallback` would need to keep track of the match var types separately from the current `Results`.

We could merge `PatternMatchingTypeCheckCompletionCallback` and `UnresolvedMemberTypeCheckCompletionCallback` so that `UnresolvedMemberTypeCheckCompletionCallback` also has `ExprPattern *PatternMatching` and `SmallVector<Type, 1> EnumPatternTypes` members, which can be `nullptr`/empty if we are not completing in a pattern matching location. I’m not sure if I like that better though… WDYT?

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@rintaro helpfully noted that the previous implementation didn’t support nested enum pattern matching, e.g. in the following, we weren’t suggesting `foo`.
```
enum PlainEnum {
  case foo(Int)
}
enum WithNestedEnum {
  case bar(PlainEnum)
}
func completeInNestedPatternMatching(x: WithNestedEnum) {
  switch x {
  case .bar(.#^COMPLETE^#)
}
```

I remodeled the code completion to not be a custom completion kind anymore (like @rintaro suggested earlier). Instead, we discover if the code completion is inside a pattern matching location (by looking if it’s an argument to an implicit `~=` operator where the second variable is named `$match`. If this is the case, we also suggest all enum members of `$match`’s type, like we were previously doing in the custom completion kind.

ahoppen commented:
@swift-ci Please test

ahoppen commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -4665,6 +4675,21 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
 
+  /// Complete all enum members declared on \p T.
+  void getUnresolvedEnumElementCompletions(Type T) {
+    if (!T->mayHaveMembers())
```
Seems like a reasonable and easy performance improvement. 👍

ahoppen commented:
@swift-ci Please smoke test

rintaro reviewed (approved):
Implementation LGTM except a couple of comments.
Please cooperate with Hamish when landing this.

rintaro commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -4694,6 +4704,21 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
 
+  /// Complete all enum members declared on \p T.
+  void getUnresolvedEnumElementCompletions(Type T) {
+    if (!T->mayHaveMembers() || !T->getAnyNominal() || !isa<EnumDecl>(T->getAnyNominal()))
```
I imported `isa_and_nonnull` into `swift` namespace in #39078 so just
```suggestion
    if (!isa_and_nonnull<EnumDecl>(T->getAnyNominal()))
```

ahoppen replied:
Nice. I didn’t know that existed. Thanks.

ahoppen commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -4665,6 +4677,21 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
   }
 
+  /// Complete all enum members declared on \p T.
+  void getUnresolvedEnumElementCompletions(Type T) {
```
Thanks. Good catch. I changed it at some point but that rename must have gotten lost in the refactoring. Renamed it again.

ahoppen committed:
[CodeCompletion] Explicitly support enum pattern matching

Pattern matching in Swift can either be expression pattern matching by comparing two instances using the `~=` operator or using enum matching by matching the enum case and its associated types (+ tuple pattern matching, but that’s not relevant here). We currenlty only consider the expression pattern matching case for code completion. To provide enum pattern matching results, we thus need to have a `~=` operator between the code completion token and the match expression

For example, when we are completing

```swift
enum MyEnum {
  case myCase(String)
}

switch x {
case .#^COMPLETE^#
}
```
then we are looking up all overloads of `~=` and try to match it to the call arguments `(<Code Completion Type>, MyEnum)`.
The way we currently get `#^COMPLETE^#` to offer members of `MyEnum`, is that we are trying to make sure that the `~=<T: Equatable>(T, T)` operator defined in the standard library is the best solution even though it has fixes associated with it. For that we need to carefully make sure to ignore other, more favourable overloads of `~=` in `filterSolutions` so that `~=<T: Equatable>(T, T)` has the best score.

This poses several problems:
- If the user defines a custom overload of `~=` that we don't prune when filtering solutions (e.g. `func ~=(pattern: OtherType, value: MyEnum) -> Bool`), it gets a better score than `~=<T: Equatable>(T, T)` and thus we only offer members of `OtherType` instead of members from `MyEnum`
- We are also suggesting static members of `MyEnum`, even though we can't pattern match them due to the lack of the `~=` operator.

If we detect that the completion expression is in a pattern matching position, also suggests all enum members of the matched type. This allows us to remove the hack which deliberately ignores certain overloads of `~=` since we no longer rely on `~=<T: Equatable>(T, T)`. It thus provides correct results in both of the above cases.

Fixes rdar://77263334 [SR-14547]

ahoppen commented:
@swift-ci Please smoke test
